confluent_kafka/consumer.py

```diff
@@ -0,0 +1,145 @@
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict.
+    """
+
+    __slots__ = ["_key_deserializer", "_value_deserializer"]
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_deserializer=None, value_deserializer=None,
+                 on_commit=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
+        if not on_commit:
+            on_commit = conf.get('on_commit', None)
+
+        if not stats_cb:
+            stats_cb = conf.get('stats_cb', None)
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+        super(Consumer, self).__init__(conf, on_commit=on_commit, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def poll(self, timeout=-1.0, key_deserializer=None, value_deserializer=None):
+        """
+            Consumes a message, triggers callbacks, returns an event.
+
+            The application must check the returned Message object’s Message.error() method to distinguish
+            between proper messages, an error(see error().code() for specifics), or an event.
+
+            :param float timeout:  Maximum time in seconds to block waiting for message, event or callback.
+            :param func key_deserializer(topic, key): Converts message key bytes to object.
+                **note** deserializers are responsible for handling NULL keys
+            :param func value_deserializer(topic, value): Converts message value bytes to object.
+                **note** deserializers are responsible for handling NULL values
+            :returns: A confluent_kafka.Message or None on timeout.
+            :raises RuntimeError: If called on a closed consumer.
+        """
+
+        msg = super(Consumer, self).poll(timeout)
+
+        if not msg or msg.error():
+            return msg
+
+        topic = msg.topic()
+
+        # parameter overrides take precedence over instance functions
```

Comment:
the below code silently omits to set anything for message key and value if the deserializers aren't specified - that's not a good idea. we could either throw an exception of pass the value straight through. the latter seems better on first instinct, but that requires further thought.

Reply:
If no serializers the message contents remain in place unaltered.

Reply:
Note the original consumer/producer are no longer exposed directly. If you run the existing tests you will see everything works as if these wrappers were never written.

Reply:
oh right, i miss-read this, and also assumed you did the same thing later...